libqtile/widget/statusnotifier.py

```diff
@@ -0,0 +1,897 @@
+            path=path,
+            display_menu_callback=self.display_menu_callback,
+            icon_theme=self.icon_theme
+        )
+        item.on_icon_changed = self.widget.bar.draw
+        if item not in self.items:
+            task = asyncio.create_task(item.start())
+            task.add_done_callback(
+                partial(self.item_added, item, service)
+            )
+
+    def remove_item(self, interface):
+        # Check if the interface is in out list of items and, if so,
+        # remove it and redraw the bar
+        if interface in self.items:
+            self.items.remove(interface)
+            self.widget.bar.draw()
+
+
+host = StatusNotifierHost()  # noqa: E303
+
+
+class StatusNotifier(base._Widget):
+    """
+    A 'system tray' widget using the freedesktop StatusNotifierItem
+    specification.
+
+    Letf-clicking an icon will trigger an activate event.
+    Right-clicking an icon will display a context menu where provided
+    by the application.
+    """
+
+    orientations = base.ORIENTATION_BOTH
+
+    defaults = [
+        ('icon_size', 16, 'Icon width'),
+        ('icon_theme', None, 'Name of theme to use for app icons'),
+        ('padding', 5, 'Padding between icons'),
+        ('font', 'sans', 'Font for menu text')
+    ]
+
+    def __init__(self, **config):
+        base._Widget.__init__(self, bar.CALCULATED, **config)
+        self.add_defaults(StatusNotifier.defaults)
+        self.add_callbacks(
+            {
+                "Button1": self.activate,
+            }
+        )
+
+    @property
+    def available_icons(self):
+        return [item for item in self.host.items if item.has_icons]
+
+    def calculate_length(self):
+        if not self.host.items:
+            return 0
+
+        return (
+            len(self.available_icons) * (self.icon_size + self.padding)
+            + self.padding
+        )
+
+    def _configure(self, qtile, bar):
+        if self.configured:
+            return
+
+        self.host = host
+        if has_xdg and self.icon_theme:
+            self.host.icon_theme = self.icon_theme
+
+        # This is called last as it starts timers including _config_async.
+        base._Widget._configure(self, qtile, bar)
+
+    async def _config_async(self):
+        await self.host.start(self)
+
+    def find_icon_at_pos(self, x, y):
+        """returns StatusNotifierItem object for icon in given position"""
+        xoffset = self.padding
+        if x < xoffset:
+            return None
+
+        for icon in self.available_icons:
+            xoffset += self.icon_size
+            if x < xoffset:
+                return icon
+            xoffset += self.padding
+
+        return None
+
+    def button_press(self, x, y, button):
+        icon = self.find_icon_at_pos(x, y)
+        if not icon:
+            return
+
+        name = 'Button{0}'.format(button)
+        if name in self.mouse_callbacks:
+            self.mouse_callbacks[name](icon)
```

Comment:
Can we keep the arguments passed to the mouse callbacks consistent with other widgets? The standard is not to have any arguments, and this inconsistency will confused anyone who adds a callback to this widget. Alternatively the `icon` could be saved to `self.clicked_icon`, or like the `GroupBox` widget the `x` value could be saved to `self.click` so that the calbacks can use `self.find_icon_at_pos(x)` if they need to - meaning that it can be avoided if it isn't needed for this click.

Reply:
Yes. Makes sense.